### PR TITLE
#261  Document catalog terms with classes Arch

### LINKED ISSUE #261: models.py:143-146: Document terms Product, Position and...
The puzzle `257-457f55e0` from #257 has to be resolved: 

https://github.com/fidals/refarm-site/blob/b00cfe6c36ef72ffd23f5a6ac54d2d7b23e2dd44/catalog/models.py#L143-L146

The puzzle was created by duker33 on 11-Feb-19. 

  

If you have any technical questions, don't ask me, submit new tickets instead. The task will be \"done\" when the problem is fixed and the text of the puzzle is _removed_ from the source code. Here is more about [PDD](http://www.yegor256.com/2009/03/04/pdd.html) and [about me](http://www.yegor256.com/2017/04/05/pdd-in-action.html).

### PR DESCRIPTION
**Pull request by duker33**
Create Position class

Closes #261

### CODE REVIEW

File: catalog/models.py [hunk 1]
```diff
@@ -142,25 +142,25 @@ def tagged(self, tags):
...
-class AbstractProduct(models.Model, AdminTreeDisplayMixin):
+# @todo #261:60m  Use the new catalog arch at sites.
+class Position(models.Model):
     """
-    Product model.
-    Defines basic functionality and primitives for Product in typical e-shop.
-    Has n:1 relation with Category.
+    The smallest unique catalog item.
+
+    By catalog semantics position has is the description of phisical product variation.
```
Comment (artemiy312) [inline catalog/models.py]:
```suggestion
    In catalog semantics, position is a description of physical variation of a product.
```

File: catalog/models.py [hunk 2]
```diff
@@ -142,25 +142,25 @@ def tagged(self, tags):
...
+    That means that only position can have price, vendor code,
```
Comment (artemiy312) [inline catalog/models.py]:
```suggestion
    Only position can have price, vendor code,
```

File: catalog/models.py [hunk 3]
```diff
@@ -142,25 +142,25 @@ def tagged(self, tags):
...
+    be or not in stock and be presented as cart item.
+    Example:
+        1. If some shop sells only one iphone X variation,
```
Comment (artemiy312) [inline catalog/models.py]:
```suggestion
        1. If a shop sells only one iphone X variation,
```

File: catalog/models.py [hunk 4]
```diff
@@ -142,25 +142,25 @@ def tagged(self, tags):
...
+        this iphone X will be both Position and Product item.
+
+        2. If some shop sells two iphone X variations: red and black,
+        iphone X will Product item,
+        iphone X will be both and Position and Option item.
```
Comment (artemiy312) [inline catalog/models.py]: This is little bit unclear, did you mean here `iphone X red/black`?

File: catalog/models.py [hunk 5]
```diff
@@ -178,6 +178,34 @@ class Meta:
         verbose_name=_('is popular'),
     )
 
+
+class AbstractOption(Position):
+    """
+    Product variation.
+
+    For example for product "iphone X" position can be "iphone X red".
```
Comment (artemiy312) [inline catalog/models.py]:
```suggestion
    "iphone X red" can be an option of "iphone X"  product  
```

File: catalog/models.py [hunk 6]
```diff
@@ -142,25 +142,25 @@ def tagged(self, tags):
         return self.get_queryset().tagged(tags)
 
 
-# @todo #257:30m  Document terms Product, Position and Option
-#  Seems the best way is to do it with subclassing.
-#  Such documentation will clarify types and will be obvious for programmers.
-#  But check the docs too. May be some of this terms are mentioned there.
-class AbstractProduct(models.Model, AdminTreeDisplayMixin):
+# @todo #261:60m  Use the new catalog arch at sites.
+class Position(models.Model):
```
Comment (artemiy312) [inline catalog/models.py]: Let's add here `Abstract` prefix or drop it from other abstract models

File: catalog/models.py (hunk 5 above)
Comment (duker33) [reply catalog/models.py]: i'll still leave "For example" words and apply another sentence

File: catalog/models.py (hunk 6 above)
Comment (duker33) [reply catalog/models.py]: i'd prefer to add it

File: catalog/models.py (hunk 1 above)
Comment (duker33) [reply catalog/models.py]: > of phisical product variation.

i'll leave this and will fix all another

**Pull request merged**